File: product/ci/addon/testintelligence/runner.go
// Copyright 2021 Harness Inc. All rights reserved.
// Use of this source code is governed by the PolyForm Free Trial 1.0.0 license
// that can be found in the licenses directory at the root of this repository, also available at
// https://polyformproject.org/wp-content/uploads/2020/05/PolyForm-Free-Trial-1.0.0.txt.

// TestRunner provides the test command and java agent config
// which is needed to run test intelligence.
package testintelligence

import (
	"context"

	"github.com/harness/ti-client/types"
)

//go:generate mockgen -source runner.go -package=testintelligence -destination mocks/runner_mock.go TestRunner
type TestRunner interface {
	// GetCmd gets the command which needs to be executed to run only the specified tests.
	// tests: list of selected tests which need to be executed
	// agentConfigPath: path to the java agent config. This needs to be added to the
	// command if instrumentation is required.
	// ignoreInstr: instrumentation might not be required in some cases like manual executions
	// runAll: if there was any issue in figuring out which tests to run, this parameter is set as true
	GetCmd(ctx context.Context, tests []types.RunnableTest, userArgs, agentConfigPath string, ignoreInstr, runAll bool) (string, error)

	// AutoDetectPackages detects the list of packages to be instrumented.
	// Return an error if we could not detect or if it's unimplemented.
	AutoDetectPackages() ([]string, error)

	// AutoDetectTests detects the list of tests in the workspace
	// Return an error if we could not detect or if it's unimplemented
	AutoDetectTests(ctx context.Context, testGlobs []string) ([]types.RunnableTest, error)
}

File: 260-delegate/src/main/java/io/harness/delegate/service/runners/itfc/Runner.java
/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Free Trial 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/05/PolyForm-Free-Trial-1.0.0.txt.
 */

package io.harness.delegate.service.runners.itfc;

import io.harness.delegate.core.beans.InputData;
import io.harness.delegate.service.handlermapping.context.Context;

public interface Runner {
  /**
   * Set up execution infrastructure
   * @param infraId  The designated infra id for the execution infra to be created
   * @param infra  Input data for infra config
   * @param context  Context of this task, including delegate information, task id, or decrypted secrets
   */
  void init(String infraId, InputData infra, Context context);
  /**
   * Execute a task
   * @param infraId  The execution infra to execute the task on
   * @param infra  Input data for the task
   * @param context  Context of this task, including delegate information, task id, or decrypted secrets
   */
  void execute(String infraId, InputData tasks, Context context);
  /**
   * Clean up an execution infra
   * @param infraId  The execution infra to clean up
   * @param context  Context of this task, including delegate information, task id, or decrypted secrets
   */
  void cleanup(String infraId, Context context);
}
